include/rect.h

```diff
@@ -29,33 +31,79 @@ namespace DosBox {
 struct Rect {
 	constexpr Rect() = default;
 
-	constexpr Rect(const float _w, const float _h)
-	        : x(0.0f),
-	          y(0.0f),
-	          w(_w),
-	          h(_h)
-	{}
-
-	constexpr Rect(const int _w, const int _h)
-	        : x(static_cast<float>(0.0f)),
-	          y(static_cast<float>(0.0f)),
-	          w(static_cast<float>(_w)),
-	          h(static_cast<float>(_h))
-	{}
 	constexpr Rect(const float _x, const float _y, const float _w, const float _h)
 	        : x(_x),
 	          y(_y),
 	          w(_w),
 	          h(_h)
+	{
+		assert(w >= 0.0f);
+		assert(h >= 0.0f);
+	}
+
+	constexpr Rect(const float _w, const float _h)
+	        : Rect(0.0f, 0.0f, _w, _h)
 	{}
 
 	constexpr Rect(const int _x, const int _y, const int _w, const int _h)
-	        : x(static_cast<float>(_x)),
-	          y(static_cast<float>(_y)),
-	          w(static_cast<float>(_w)),
-	          h(static_cast<float>(_h))
+	        : Rect(static_cast<float>(_x), static_cast<float>(_y),
+	               static_cast<float>(_w), static_cast<float>(_h))
 	{}
 
+	constexpr Rect(const int _w, const int _h) : Rect(0, 0, _w, _h) {}
+
+	constexpr float x1() const
+	{
+		return x;
+	}
+
+	constexpr float y1() const
+	{
+		return y;
+	}
+
+	constexpr float x2() const
+	{
+		return x + w;
+	}
+
+	constexpr float y2() const
+	{
+		return y + h;
+	}
+
+	constexpr float cx() const
+	{
+		return x + w / 2;
+	}
+
+	constexpr float cy() const
+	{
+		return y + h / 2;
+	}
+
+	constexpr Rect Scale(const float s) const
+	{
+		return {x * s, y * s, w * s, h * s};
+	}
+
+	constexpr Rect Translate(const float dx, const float dy) const
+	{
+		return {x + dx, y + dy, w, h};
+	}
```

Comment:
These are member functions, so they really should modify the object itself. 

So I think these members should modify the object and then return a reference to itself, with a signature like:

```c++
Rect& Translate(const float dx, const float dy) {
   modify itself
   ...
   return *this;
}
```

This will allow for chaining: `my_rect.Scale(0.5).Translate(1, 1);`

To illustrate: imagine calling `my_vector.push_back(item)`, but it not actually adding it to `my_vector` and instead only the returned result had it added.

The pure functions have a place, but normally they'd take in the object being operated on in the args, and would be stand-alone functions.

Reply:
> These are member functions, so they really should modify the object itself.

As a user of functional languages on a daily basis at work, I disagree 😄 Those have sizable standard libraries and almost nothing mutates; all class methods return copies. Then there are some mutating variants, but their use is discouraged—you have to actively ask for it, never mutates by default.

But ok, we're in C++. I think the compiler would optimise it anyway, but I don't care much, happy to mutate. Btw, these rects won't be used in any high-performance code.

Reply:
It's how all STL C++ objects work when you call their member functions.

```c++
string s = 'hello';
s.append(' world');
```

We expect that `s` contains `hello world`.

If it behaved like the above, we would need:

```c++
string s = 'hello';
const auto new_s = s.append(' world');
```

- `s` would still contain `hello`
- `new_s` would have `hello world`.

Reply:
Well, that's STL 😄  Yep, that's exactly how strings work in "immutable by default" languages. All strings are immutable. You need to ask for special `MutableStrings` if you want to muck around with the contents of a strings in place. It's actually nice because if you use mutable objects in hash maps that can lead to obscure bugs if you mutate them later... (been there in Java 😅 )

Ok, I've rewritten it to mutate. I kinda liked the first form better but I can live with it 😄 

```cpp
const auto viewport_px = limit_rect.Scale(sdl.desktop.dpi_scale);

return canvas_px.Contains(viewport_px)
			 ? viewport_px
			 : viewport_px.Intersect(canvas_px);

```

```cpp
auto viewport_px = limit_rect;
viewport_px.Scale(sdl.desktop.dpi_scale);

if (!canvas_px.Contains(viewport_px)) {
	viewport_px.Intersect(canvas_px);
}

return viewport_px;
```

Reply:
Okay, I've done the always return self-reference trick, plus also introduced a `Copy()` method so we can copy & transform & assign one rect to another const rect. It's not so bad, although more error-prone than the pure version. But so is STL, so ThisIsFine(tm) 😆